### ISSUE TITLE
Error: undefined method `before_action' for Trestle::Resource::Controller:Module

### ISSUE THREAD

**Issue by basharabdullah** [richardvenneman/trestle-active_storage#58]
Hi,

I just updated my Trestle gems to latest to fix some bug, and I started getting this error below, apparently related to active_storage.

```
1:25:45 PM web.1 |  Puma starting in single mode...
1:25:45 PM web.1 |  * Version 4.3.3 (ruby 2.6.2-p47), codename: Mysterious Traveller
1:25:45 PM web.1 |  * Min threads: 5, max threads: 5
1:25:45 PM web.1 |  * Environment: development
1:25:51 PM web.1 |  ! Unable to load application: NoMethodError: undefined method `before_action' for Trestle::Resource::Controller:Module
1:25:51 PM web.1 |  bundler: failed to load command: puma (/Users/bashar3a/.rbenv/versions/2.6.2/bin/puma)
1:25:51 PM web.1 |  NoMethodError: undefined method `before_action' for Trestle::Resource::Controller:Module
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/trestle-active_storage-2.2.1/lib/trestle/active_storage/controller_concern.rb:7:in `block in <module:ControllerConcern>'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.3/lib/active_support/concern.rb:122:in `class_eval'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/activesupport-6.0.3/lib/active_support/concern.rb:122:in `append_features'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/trestle-active_storage-2.2.1/lib/trestle/active_storage/engine.rb:9:in `include'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/trestle-active_storage-2.2.1/lib/trestle/active_storage/engine.rb:9:in `block in <class:Engine>'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/railties-6.0.3/lib/rails/initializable.rb:32:in `instance_exec'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/railties-6.0.3/lib/rails/initializable.rb:32:in `run'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/railties-6.0.3/lib/rails/initializable.rb:61:in `block in run_initializers'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:228:in `block in tsort_each'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:431:in `each_strongly_connected_component_from'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:349:in `block in each_strongly_connected_component'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:347:in `each'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:347:in `call'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:347:in `each_strongly_connected_component'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:226:in `tsort_each'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/2.6.0/tsort.rb:205:in `tsort_each'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/railties-6.0.3/lib/rails/initializable.rb:60:in `run_initializers'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/railties-6.0.3/lib/rails/application.rb:363:in `initialize!'
1:25:51 PM web.1 |    /Users/bashar3a/rails-projects/example/config/environment.rb:5:in `<top (required)>'
1:25:51 PM web.1 |    config.ru:3:in `require_relative'
1:25:51 PM web.1 |    config.ru:3:in `block in <main>'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rack-2.2.2/lib/rack/builder.rb:116:in `eval'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rack-2.2.2/lib/rack/builder.rb:116:in `new_from_string'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rack-2.2.2/lib/rack/builder.rb:105:in `load_file'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/rack-2.2.2/lib/rack/builder.rb:66:in `parse_file'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/puma-4.3.3/lib/puma/configuration.rb:321:in `load_rackup'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/puma-4.3.3/lib/puma/configuration.rb:246:in `app'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/puma-4.3.3/lib/puma/runner.rb:155:in `load_and_bind'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/puma-4.3.3/lib/puma/single.rb:98:in `run'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/puma-4.3.3/lib/puma/launcher.rb:172:in `run'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/puma-4.3.3/lib/puma/cli.rb:80:in `run'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/lib/ruby/gems/2.6.0/gems/puma-4.3.3/bin/puma:10:in `<top (required)>'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/bin/puma:23:in `load'
1:25:51 PM web.1 |    /Users/bashar3a/.rbenv/versions/2.6.2/bin/puma:23:in `<top (required)>'
```

My gemfile

```
gem 'trestle', '=0.9.3'
gem 'trestle-auth', '=0.4.0'
gem 'trestle-search', '=0.4.2'
gem 'trestle-tinymce', '=0.2.1'
gem 'trestle-active_storage', '=2.2.1'
```

Any idea? Or is this more related to a bug in Trestle itself?

Thanks

**Comment by kxu1988** [comment]
I changed to 0.9.2,and it is ok

**Comment by ledermann** [comment]
Same here, seems to be caused by Zeitwerk fixes in https://github.com/TrestleAdmin/trestle/commit/9b2bb96894088e278fb43d62f880171966c1b1f8

Upgrading Trestle to 0.9.3 is currently not possible while using this gem.

**Comment by basharabdullah** [comment]
@kxu1988 yes seems I will revert back. I needed the upgrade for another fix actually.

@ledermann thanks for confirming. Will wait then

**Comment by McRipper** [comment]
I've fixed it in the pull request #56 that has been merged to master, it needs to be tested, you can reference master in Gemfile.

**Comment by richardvenneman** [comment]
Sorry about that! Can you please try to use the gem from master (which contains @McRipper's fix) and let me know if that solves your issue? I'll then cut a new gem version:

```ruby
gem "trestle-active_storage", github: "richardvenneman/trestle-active_storage"
```

**Comment by gnclmorais** [comment]
Not sure if entirely related to the [latest version on master](https://github.com/richardvenneman/trestle-active_storage/issues/58#issuecomment-630699186) right now, but I’m running into this on Heroku and locally:
```
2020-05-24T21:03:48.905959+00:00 app[web.1]: I, [2020-05-24T21:03:48.905879 #4]  INFO -- : [110013ab-5fa2-4082-b6a0-59b88faa2380] Completed 500 Internal Server Error in 21ms (ActiveRecord: 1.7ms | Allocations: 6626)
2020-05-24T21:03:48.907532+00:00 app[web.1]: F, [2020-05-24T21:03:48.907459 #4] FATAL -- : [110013ab-5fa2-4082-b6a0-59b88faa2380]
2020-05-24T21:03:48.907534+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380] ActionView::Template::Error (undefined method `delete_image' for #<Advert:0x000055f1538f1b88>):
2020-05-24T21:03:48.907534+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]     16:       <div><small><%= attachment.blob.filename %></small></div>
2020-05-24T21:03:48.907535+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]     17:     </div>
2020-05-24T21:03:48.907536+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]     18:
2020-05-24T21:03:48.907536+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]     19:     <%= builder.check_box("delete_#{field_name}", label: "Delete #{field_name}") %>
2020-05-24T21:03:48.907537+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]     20:   <% end %>
2020-05-24T21:03:48.907537+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]     21:
2020-05-24T21:03:48.907538+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]     22:   <div class="progress" style="display: none">
2020-05-24T21:03:48.907538+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380]
2020-05-24T21:03:48.907538+00:00 app[web.1]: [110013ab-5fa2-4082-b6a0-59b88faa2380] app/admin/adverts_admin.rb:29:in `block (2 levels) in <main>'
```
![image](https://user-images.githubusercontent.com/385232/82764975-e0681300-9e0a-11ea-9184-705392ba5fe3.png)

**Comment by ledermann** [comment]
I stumbled upon the same issue as @gnclmorais told. The attr_accessor `delete_{fieldname}` does not exist. The reason is that the method [define_attachment_accessors](https://github.com/richardvenneman/trestle-active_storage/blob/master/lib/trestle/active_storage/controller_concern.rb#L12) is not executed. in the [ControllerConcern](https://github.com/richardvenneman/trestle-active_storage/blob/master/lib/trestle/active_storage/controller_concern.rb).

@richardvenneman, @McRipper: Currently, this gem is not ready to release. It seems there is still a class loading issue.

**Comment by McRipper** [comment]
Trestle changed the structure to support new Rails autoloading, it could be that we need to change the folder structure too to support the new release. I will try to take a look ASAP.

**Comment by HatsuMora** [comment]
I ran into both issues, I have time to help if someone guides me in the right way.

**Comment by McRipper** [comment]
Just waiting @richardvenneman to release a new version, in the meantime you can test by pointing at the git repo.

**Comment by richardvenneman** [comment]
I've just released `trestle-active_storage` `3.0.0` which contains @McRipper efforts to fix this issue. Thanks again!
